### PR TITLE
Dg diffusion

### PR DESCRIPTION
**Pull request by arotem3**
__Summary:__ partial assembly implementation of the `DGDiffusionIntegrator`.
Recall, the bilinear form of DG diffusion is:

`B(u, v) := - < {(Q grad(u)).n}, [v] > + sigma < [u], {(Q grad(v)).n} > + kappa < {h^{-1} Q} [u], [v] >`

- `DGDiffusionIntegrator` now supports PA in 2D and 3D when `vdim == 1` for tensor product elements. The `Mult` method was optimized for GPU use (specifically CUDA).
- Added `L2NormalDerivativeFaceRestriction` which computes the normal derivative in _reference_ coordinates on the collocation points on edges. The normal derivative in physical coordinates can be reconstructed using the tangential derivatives in reference coordinates which are computed directly from the face values provided by `L2FaceRestriction`. The derivative is computed in reference coordinates to match the approach of the other face restrictions in MFEM (the normal derivative in physical coordinates cannot be represented exactly on the collocation points, hence we return the values in reference coordinates which can).
- example 14 now includes PA as an option.
- Added tests "PA DG Diffusion" and "Parallel PA DG Diffusion".

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3904](https://github.com/mfem/mfem/pull/3904) | @arotem3 | @pazner | @camierjs + @v-dobrev + @YohannDudouit + @artv3 | 10/2/23 | 4/17/24 | 4/18/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by pazner** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/bilininteg.cpp [hunk 1]
```diff
@@ -3458,17 +3464,7 @@ void DGDiffusionIntegrator::AssembleFaceMatrix(
...
-      if (ndof2)
-      {
-         order = 2*max(el1.GetOrder(), el2.GetOrder());
-      }
-      else
-      {
-         order = 2*el1.GetOrder();
-      }
-      ir = &IntRules.Get(Trans.GetGeometryType(), order);
+      ir = &GetRule(el1.GetOrder(), Trans);
```
Comment (termi-official) [inline fem/bilininteg.cpp:3467]: Is there an easy way to preserve this behavior?

Comment (pazner) [reply fem/bilininteg.cpp:3467]: You mean using the max polynomial degree of neighboring elements?

Comment (termi-official) [reply fem/bilininteg.cpp:3467]: Indeed. I think keeping the maximum polynomial degree per default might still be of interest for p and hp adaptivity (in the future).

Comment (pazner) [reply fem/bilininteg.cpp:3467]: Yes, it should be easy to replace `el1.GetOrder()` with the max (if there is a second element, following the logic of the deleted code). cc: @arotem3

**Review by YohannDudouit — COMMENTED** [review]
Very good work overall! Just a few minor comments, once these are resolved, I'll run and approve.

File: fem/bilininteg.hpp [hunk 2]
```diff
@@ -266,6 +266,11 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
                                     Vector &flux, Vector *d_energy = NULL)
    { return 0.0; }
 
+   virtual bool RequiresFaceNormalDerivatives() const { return false; }
+
+   virtual void AddMultPAFaceNormalDerivatives(const Vector &x, const Vector &dxdn,
```
[the remark this answers is not in the document]
Comment (arotem3) [reply fem/bilininteg.hpp:298]: please see*

File: fem/bilininteg.cpp [hunk 3]
```diff
@@ -3458,17 +3464,9 @@ void DGDiffusionIntegrator::AssembleFaceMatrix(
...
-         order = 2*max(el1.GetOrder(), el2.GetOrder());
-      }
-      else
-      {
-         order = 2*el1.GetOrder();
-      }
-      ir = &IntRules.Get(Trans.GetGeometryType(), order);
+      const int order = (ndof2) ? max(el1.GetOrder(),
+                                      el2.GetOrder()) : el1.GetOrder();
+      ir = &GetRule(order, Trans);
```
Comment (YohannDudouit) [inline fem/bilininteg.cpp:3469]: What is motivating this change? Also, I think `GetRule` should be in `bilininteg.cpp`, it is currently very hard to locate!

Comment (YohannDudouit) [reply fem/bilininteg.cpp:3469]: ping @pazner @arotem3

Comment (arotem3) [reply fem/bilininteg.cpp:3469]: This change was made in reference to an earlier comment by @termi-official in consideration for possible p (or hp) adaptivity in the future. Does this answer the question @YohannDudouit ?

Comment (termi-official) [reply fem/bilininteg.cpp:3469]: IIRC then we have some HP baseline already functional, see https://github.com/mfem/mfem/pull/1423

File: fem/bilininteg.hpp [hunk 4]
```diff
@@ -266,6 +266,11 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
                                     Vector &flux, Vector *d_energy = NULL)
    { return 0.0; }
 
+   virtual bool RequiresFaceNormalDerivatives() const { return false; }
```
[the remark this answers is not in the document]
Comment (arotem3) [reply fem/bilininteg.hpp:279]: All docs updated in: https://github.com/mfem/mfem/pull/3904/commits/e5d74380d0215c935502eb41fdc1f088a5941dc2

[the remark this answers is not in the document]
Comment (arotem3) [reply fem/bilininteg.hpp:279]: also in https://github.com/mfem/mfem/pull/3904/commits/e8ec967feefb6231590c0805217b00a4d14518ce

[the remark this answers is not in the document]
Comment (arotem3) [reply fem/bilininteg.hpp:279]: and https://github.com/mfem/mfem/pull/3904/commits/cf5fd2b3e1bb1823a9d68972e6e093a9b991cfce

File: fem/bilininteg.cpp (hunk 3 above)
Comment (pazner) [reply fem/bilininteg.cpp:3469]: As of 98cfc1fd6, the exact same quadrature rules should be used before and after this PR (just factored out to the `GetRule` member function)

**Review by pazner — APPROVED** [review]
Great work, @arotem3!

@YohannDudouit, I think your comments should be addressed now.

@artv3, @camierjs, @v-dobrev, can you take a look when you get a chance?

**Review by YohannDudouit — APPROVED** [review]
Very nice work, and very clean overall.

I really like the use of stack allocation instead of heap allocation in `ex14`, this makes the example look a lot less amateurish from an HPC perspective, and also results in much more idiomatic C++.

**Comment by pazner** [comment]
Thanks for the review @YohannDudouit!

@artv3, @camierjs, @v-dobrev, could you take a look when you get a chance? Thanks!

**Comment by artv3** [comment]
Can we update with master?

**Review by artv3 — APPROVED** [review]
Works for us!

**Comment by tzanio** [comment]
Now that PR https://github.com/mfem/mfem/pull/3922 has been merged in `master`, please update this PR by:

1. merging `master` in here and resolving conflicts
2. replacing use of `double` with `real_t`
3. guarding double/single-only sections with `MFEM_USE_DOUBLE` and `MFEM_USE_SINGLE`
4. making sure tests run cleanly with default and `MFEM_PRECISION=single` configurations

See https://github.com/orgs/mfem/discussions/4207 for more details and/or ask @dylan-copeland for help.

**Comment by tzanio** [comment]
> Now that PR https://github.com/mfem/mfem/pull/3922 has been merged in master, please update this PR by:
> 
> 1. merging master in here and resolving conflicts
> 2. replacing use of double with real_t
> 3. guarding double/single-only sections with MFEM_USE_DOUBLE and MFEM_USE_SINGLE
> 4. making sure tests run cleanly with default and MFEM_PRECISION=single configurations
>
> See https://github.com/orgs/mfem/discussions/4207 for more details and/or ask @dylan-copeland for help.

ping: @arotem3, @pazner

**Comment by pazner** [comment]
> > Now that PR #3922 has been merged in master, please update this PR by:
> > 
> > 1. merging master in here and resolving conflicts
> > 2. replacing use of double with real_t
> > 3. guarding double/single-only sections with MFEM_USE_DOUBLE and MFEM_USE_SINGLE
> > 4. making sure tests run cleanly with default and MFEM_PRECISION=single configurations
> > 
> > See https://github.com/orgs/mfem/discussions/4207 for more details and/or ask @dylan-copeland for help.
> 
> ping: @arotem3, @pazner

I made these changes, I think it should be OK now.

**Comment by tzanio** [comment]
Thank you, @pazner !

File: examples/ex14.cpp [hunk 5]
```diff
@@ -19,6 +19,10 @@
 //               ex14 -m ../data/amr-hex.mesh
 //               ex14 -m ../data/fichera-amr.mesh
 //
+// Device sample runs:
+//               ex14 -pa -d cuda -o 3
+//               ex14 -pa -d cuda -o 3 -m ../data/fichera.mesh -o 3
```
Comment (v-dobrev) [inline examples/ex14.cpp:24]: The arguments `-o 3` are repeated:
```suggestion
//               ex14 -pa -d cuda -o 3 -m ../data/fichera.mesh
```

File: examples/ex14.cpp [hunk 6]
```diff
@@ -62,132 +68,154 @@ int main(int argc, char *argv[])
...
    {
-      PCG(A, M, *b, x, 1, 500, 1e-12, 0.0);
+      const Operator &A = a;
+      if (sigma == -1.0)
+      {
+         CG(A, b, x, 1, 500, 1e-24, 0.0);
+      }
+      else
+      {
+         MFEM_ABORT("Not yet supported.");
```
Comment (v-dobrev) [inline examples/ex14.cpp:177]:
```suggestion
         MFEM_ABORT("The case of PA with sigma != -1 is not yet supported.");
```

File: examples/ex14.cpp [hunk 7]
```diff
@@ -62,132 +68,154 @@ int main(int argc, char *argv[])
...
       args.PrintUsage(cout);
       return 1;
    }
+   args.PrintOptions(cout);
    if (kappa < 0)
    {
       kappa = (order+1)*(order+1);
+      cout << "Using kappa = " << kappa << endl;
    }
-   args.PrintOptions(cout);
```
Comment (v-dobrev) [inline examples/ex14.cpp:78]: For consistency, it's good to make this change in `ex14p` as well. Or, if you prefer, you can revert this change.

File: fem/normal_deriv_restriction.hpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+   Array<int> elem_to_face; ///< Element-wise information array.
+   Array<int> face_to_vol; ///< maps face index to volume index
+
+public:
+   /// @brief Constructor.
+   /// @param[in] fes_ The associated FiniteElementSpace (should be L2/DG).
+   /// @param[in] ordering The ordering of L-DOFs.
+   /// @param[in] face_type_ Type of faces to compute restriction (interior or boundary).
+   L2NormalDerivativeFaceRestriction(const FiniteElementSpace &fes_,
+                                     const ElementDofOrdering ordering,
```
Comment (v-dobrev) [inline fem/normal_deriv_restriction.hpp:45]: The ordering argument comes from the `f_ordering` argument of the `L2FaceRestriction` constructor, so it has the same meaning:
```suggestion
   /// @param[in] f_ordering Request a specific face dof ordering
   /// @param[in] face_type_ Type of faces to compute restriction (interior or boundary).
   L2NormalDerivativeFaceRestriction(const FiniteElementSpace &fes_,
                                     const ElementDofOrdering f_ordering,
```

File: fem/normal_deriv_restriction.cpp [hunk 9]
```diff
@@ -0,0 +1,854 @@
...
+               f2v(p, side, f) = i + d * (j + d * k);
+            }
+         }
+      }
+   }
+}
+
+L2NormalDerivativeFaceRestriction::L2NormalDerivativeFaceRestriction(
+   const FiniteElementSpace &fes_,
+   const ElementDofOrdering ordering,
```
Comment (v-dobrev) [inline fem/normal_deriv_restriction.cpp:96]: For consistency with the declaration:
```suggestion
   const ElementDofOrdering f_ordering,
```

Comment (v-dobrev) [reply fem/normal_deriv_restriction.cpp:96]: Currently, this argument is not used. The assumption is that it is `ElementDofOrdering::LEXICOGRAPHIC`, as in `L2FaceRestriction::CheckFESpace()`. For clarity, it is good to add a comment about this and add an `MFEM_ASSERT`/`MFEM_VERIFY` that the other case is not implemented.

File: fem/normal_deriv_restriction.hpp [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+                                     const FaceType face_type_);
+
+   /// @brief Computes the normal derivatives on the @a face_type faces of the mesh.
+   /// @param[in] x The L-vector degrees of freedom.
+   /// @param[out] y The face E(like)-vector degrees of freedom of the format
+   /// (face_dofs x vdim x 2 x nf) where nf is the number of faces of type @a
+   /// face_type. The face_dofs are ordered according to @a ordering specified
```
Comment (v-dobrev) [inline fem/normal_deriv_restriction.hpp:52]:
```suggestion
   /// face_type. The face_dofs are ordered according to @a f_ordering specified
```

File: fem/bilinearform_ext.cpp [hunk 11]
```diff
@@ -558,14 +590,37 @@ void PABilinearFormExtension::Mult(const Vector &x, Vector &y) const
...
+            int_face_dYdn = 0.0;
+         }
+
          for (int i = 0; i < iFISz; ++i)
          {
-            intFaceIntegrators[i]->AddMultPA(int_face_X, int_face_Y);
+            if (intFaceIntegrators[i]->RequiresFaceNormalDerivatives())
+            {
+               intFaceIntegrators[i]->AddMultPAFaceNormalDerivatives(int_face_X, int_face_dXdn,
+                                                                     int_face_Y, int_face_dYdn);
```
Comment (v-dobrev) [inline fem/bilinearform_ext.cpp:612]: Better formatting:
```suggestion
               intFaceIntegrators[i]->AddMultPAFaceNormalDerivatives(
                  int_face_X, int_face_dXdn,
                  int_face_Y, int_face_dYdn);
```

File: fem/bilinearform_ext.cpp [hunk 12]
```diff
@@ -558,14 +590,37 @@ void PABilinearFormExtension::Mult(const Vector &x, Vector &y) const
...
+            }
+            else
+            {
+               intFaceIntegrators[i]->AddMultPA(int_face_X, int_face_Y);
+            }
          }
          int_face_restrict_lex->AddMultTransposeInPlace(int_face_Y, y);
+         if (int_face_dYdn.Size() > 0)
+         {
+            int_face_restrict_lex->NormalDerivativeAddMultTranspose(int_face_dYdn, y);
```
Comment (v-dobrev) [inline fem/bilinearform_ext.cpp:622]: Formatting:
```suggestion
            int_face_restrict_lex->NormalDerivativeAddMultTranspose(
               int_face_dYdn, y);
```

File: fem/bilininteg.cpp [hunk 13]
```diff
@@ -189,6 +189,12 @@ void BilinearFormIntegrator::AssembleTraceFaceMatrix (int elem,
               " Integrator class.");
 }
 
+void BilinearFormIntegrator::AddMultPAFaceNormalDerivatives(const Vector &x,
+                                                            const Vector &dxdn, Vector &y, Vector &dydn) const
```
Comment (v-dobrev) [inline fem/bilininteg.cpp:193]: Formatting:
```suggestion
void BilinearFormIntegrator::AddMultPAFaceNormalDerivatives(
   const Vector &x, const Vector &dxdn, Vector &y, Vector &dydn) const
```

File: fem/integ/bilininteg_dgdiffusion_pa.cpp [hunk 14]
```diff
@@ -0,0 +1,1051 @@
...
+      kernel(NF, B, Bt, G, Gt, sigma, pa_data, x, dxdn, y, dydn, D1D, Q1D);
+   }
+   else
+   {
+      MFEM_ABORT("Unsupported dimension");
+   }
+}
+
+void DGDiffusionIntegrator::AddMultPAFaceNormalDerivatives(const Vector &x,
+                                                           const Vector &dxdn, Vector &y, Vector &dydn) const
```
Comment (v-dobrev) [inline fem/integ/bilininteg_dgdiffusion_pa.cpp:1044]: Formatting:
```suggestion
void DGDiffusionIntegrator::AddMultPAFaceNormalDerivatives(
   const Vector &x, const Vector &dxdn, Vector &y, Vector &dydn) const
```

File: fem/normal_deriv_restriction.cpp [hunk 15]
```diff
@@ -0,0 +1,854 @@
...
+
+#ifdef MFEM_USE_MPI
+   std::unique_ptr<ParGridFunction> x_gf;
+   if (const auto *pfes = dynamic_cast<const ParFiniteElementSpace*>(&fes))
+   {
+      if (face_type == FaceType::Interior)
+      {
+         x_gf.reset(new ParGridFunction(const_cast<ParFiniteElementSpace*>(pfes),
+                                        const_cast<Vector&>(x), 0));
+         x_gf->ExchangeFaceNbrData();
```
Comment (v-dobrev) [inline fem/normal_deriv_restriction.cpp:329]: It looks like we do this same exchange two times for DG diffusion. Can we factor it out to avoid the repetition?

File: mesh/face_nbr_geom.cpp [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+      const int nbr_rank = mesh->GetFaceNbrRank(i);
+      const int tag = 0;
+
+      MPI_Isend(send_data_ptr + send_offsets[i],
+                send_offsets[i+1] - send_offsets[i],
+                MPI_DOUBLE, nbr_rank, tag, comm, &send_reqs[i]);
+
+      MPI_Irecv(x_shared_ptr + recv_offsets[i],
+                recv_offsets[i+1] - recv_offsets[i],
+                MPI_DOUBLE, nbr_rank, tag, comm, &recv_reqs[i]);
```
Comment (v-dobrev) [inline mesh/face_nbr_geom.cpp:119]: For single precision: `MPI_DOUBLE` -> `MPITypeMap<real_t>::mpi_type`.

File: fem/bilinearform_ext.cpp [hunk 17]
```diff
@@ -558,14 +590,37 @@ void PABilinearFormExtension::Mult(const Vector &x, Vector &y) const
    if (int_face_restrict_lex && iFISz>0)
    {
       int_face_restrict_lex->Mult(x, int_face_X);
-      if (int_face_X.Size()>0)
+      if (int_face_dXdn.Size() > 0)
```
Comment (v-dobrev) [inline fem/bilinearform_ext.cpp:593]: With empty local mesh, this is probably not the right check -- every rank needs to call `NormalDerivativeMult` because of the `ExchangeFaceNbrData` call inside.

Comment (v-dobrev) [reply fem/bilinearform_ext.cpp:593]: At least this comment seems to imply that: https://github.com/mfem/mfem/blob/4ff5eb8899976a5f389af7a9115b62ffa0974772/fem/prestriction.cpp#L319-L321

File: fem/bilinearform_ext.hpp [hunk 18]
```diff
@@ -70,10 +70,12 @@ class PABilinearFormExtension : public BilinearFormExtension
    const FiniteElementSpace *trial_fes, *test_fes; // Not owned
    /// Attributes of all mesh elements.
    Array<int> elem_attributes, bdr_attributes;
-   mutable Vector tmp_evec; // Work array
+   mutable Vector tmp_evec; // Work arrays
```
Comment (v-dobrev) [inline fem/bilinearform_ext.hpp:73]: Why plural, "arrays"?

File: fem/bilininteg.hpp [hunk 19]
```diff
@@ -266,6 +266,39 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
...
+                AddMultPAFaceNormalDerivatives(...) should be invoked in place
+                of AddMultPA(...) and L2NormalDerivativeFaceRestriction should
+                be used to compute the normal derivatives. This is used for some
+                DG integrators, for example DGDiffusionIntegrator.
+
+       @returns whether normal derivatives appear in the bilinear form.
+   */
+   virtual bool RequiresFaceNormalDerivatives() const { return false; }
+
+   /// Method for partially assembled transposed action.
```
Comment (v-dobrev) [inline fem/bilininteg.hpp:282]: Why "transpose" -- isn't this simply the action?

File: fem/bilininteg.hpp [hunk 20]
```diff
@@ -3241,6 +3281,24 @@ class DGDiffusionIntegrator : public BilinearFormIntegrator
                                    const FiniteElement &el2,
                                    FaceElementTransformations &Trans,
                                    DenseMatrix &elmat);
+
+   virtual bool RequiresFaceNormalDerivatives() const { return true; }
```
Comment (v-dobrev) [inline fem/bilininteg.hpp:3285]: Let's replace `virtual` with `override` everywhere in this class.

File: fem/bilininteg.cpp [hunk 21]
```diff
@@ -3654,6 +3652,13 @@ void DGDiffusionIntegrator::AssembleFaceMatrix(
    }
 }
 
+const IntegrationRule &DGDiffusionIntegrator::GetRule(
+   int order, FaceElementTransformations &T)
+{
+   // order is typically the maximum of the order of the left and right elements
+   // neighboring the given face.
+   return irs.Get(T.GetGeometryType(), 2*order);
```
Comment (v-dobrev) [inline fem/bilininteg.cpp:3660]: The variable `irs` uses `Quadrature1D::GaussLobatto` -- this changes the current behavior for non-PA and I don't think that is necessary.

For PA, the use of `GaussLobatto` quadrature seems intentional and maybe necessary. If that's the case, I'd suggest defining separate methods: `GetRule` and `GetRulePA` -- this first used for non-PA, the second for PA.

File: fem/restriction.hpp [hunk 22]
```diff
@@ -487,34 +531,51 @@ class L2FaceRestriction : public FaceRestriction
    virtual void AddFaceMatricesToElementMatrices(const Vector &fea_data,
                                                  Vector &ea_data) const;
 
+   /** @brief Scatter the degrees of freedom, i.e. goes from L-Vector to
+      face E-Vector.
+
+      @param[in]  x The L-vector degrees of freedom.
+      @param[out] y The face E-Vector degrees of freedom with the given format:
+                    if L2FacesValues::DoubleValued (face_dofs x vdim x 2 x nf)
+                    if L2FacesValues::SingleValued (face_dofs x vdim x nf)
```
Comment (v-dobrev) [inline fem/restriction.hpp:540]: Currently, this method does not do anything different when `m == L2FacesValues::SingleValued` compared to `m == L2FacesValues::DoubleValued` -- `m` is ignored. Is this comment meant to describe future intended behavior? If that's the case, let's add `MFEM_VERIFY` that `m == L2FacesValues::DoubleValued` with appropriate error message. If the intent (now and in the future) is to not use `m` then let's fix this documentation.

File: examples/ex14.cpp (hunk 7 above)
Comment (pazner) [reply examples/ex14.cpp]: Reverted in d42dd9c0a

File: fem/bilinearform_ext.hpp (hunk 18 above)
Comment (pazner) [reply fem/bilinearform_ext.hpp:73]: Because in some previous draft of the code there were more than one :)
Fixed in 1bc798642

File: fem/bilininteg.hpp (hunk 20 above)
Comment (pazner) [reply fem/bilininteg.hpp:3285]: Good idea, done in 16b3b6854.

File: fem/bilininteg.hpp (hunk 19 above)
Comment (pazner) [reply fem/bilininteg.hpp:282]: Copy & paste bug. Fixed in 558fd4598.

File: mesh/face_nbr_geom.cpp (hunk 16 above)
Comment (pazner) [reply mesh/face_nbr_geom.cpp:119]: Good catch. Fixed in f613067c3.

File: fem/integ/bilininteg_dgdiffusion_pa.cpp (hunk 14 above)
Comment (pazner) [reply fem/integ/bilininteg_dgdiffusion_pa.cpp:1044]: Fixed in 9f2668b29.

File: fem/bilininteg.cpp (hunk 13 above)
Comment (pazner) [reply fem/bilininteg.cpp:193]: Fixed in 46eeb8602.

File: fem/normal_deriv_restriction.cpp (hunk 9 above)
Comment (pazner) [reply fem/normal_deriv_restriction.cpp:96]: Done in fd28c2809.

File: fem/restriction.hpp (hunk 22 above)
Comment (pazner) [reply fem/restriction.hpp]: I think we always want double-valued. Comment is fixed in 8c85e850c321e78df446949582091626b6fc7007.

File: fem/bilinearform_ext.cpp (hunk 17 above)
Comment (pazner) [reply fem/bilinearform_ext.cpp:593]: I think almost all of the PA examples fail with empty local mesh partitions. Perhaps this is something we can address in a separate PR?

(I do think it should be addressed, but debugging/testing will require making some fixes/changes to a lot of other parts of the code).

File: fem/normal_deriv_restriction.cpp (hunk 15 above)
Comment (pazner) [reply fem/normal_deriv_restriction.cpp:333]: You mean that it's done once in `ParL2FaceRestriction` and once in `L2NormalDerivativeFaceRestriction`?

It would be good to get rid of this repetition. Do you have an idea how to structure it?

File: fem/bilininteg.cpp (hunk 21 above)
Comment (pazner) [reply fem/bilininteg.cpp]: Yes, the use of Gauss-Lobatto quadrature for PA is currently necessary, because we need to get the element Jacobians at the face quadrature points. The way we do that is to get the mesh geometric factors at the Gauss-Lobatto points, and then extract the quadrature points that lie on each face. It is possible to extend this to work for arbitrary quadrature rules either by interpolating, or by extending the functionality of the QuadratureInterpolator, but it's not completely straightforward.

I changed the code in 1c0bcaf7c so that non-PA uses the default `IntRules`, and PA uses Gauss-Lobatto of equivalent order. This has the side-effect that the PA and non-PA versions may give different answers if the quadrature is not exact.

File: fem/bilinearform_ext.cpp (hunk 17 above)
Comment (v-dobrev) [reply fem/bilinearform_ext.cpp:593]: I'm not sure there are that many places where empty partitions are not supported.

I'll try to run a test to see what pops up.

File: fem/bilininteg.cpp (hunk 21 above)
Comment (v-dobrev) [reply fem/bilininteg.cpp]: Looks good. Thanks!

File: fem/bilinearform_ext.cpp (hunk 17 above)
Comment (v-dobrev) [reply fem/bilinearform_ext.cpp:593]: I created a branch https://github.com/mfem/mfem/tree/dg-diffusion-empty-parts (based on this branch) which seems to allow `ex14p` to run with empty partitions. I tested with the following change in `ex14p` which introduces one empty partition in the last rank:
```diff
diff --git a/examples/ex14p.cpp b/examples/ex14p.cpp
index 7edb8b0c7..49fe9ae36 100644
--- a/examples/ex14p.cpp
+++ b/examples/ex14p.cpp
@@ -162,7 +162,13 @@ int main(int argc, char *argv[])
    // 5. Define a parallel mesh by a partitioning of the serial mesh. Refine
    //    this mesh further in parallel to increase the resolution. Once the
    //    parallel mesh is defined, the serial mesh can be deleted.
-   ParMesh pmesh(MPI_COMM_WORLD, mesh);
+   int *partitioning = nullptr;
+   {
+      const int np = Mpi::WorldSize();
+      partitioning = mesh.GeneratePartitioning(np > 1 ? np-1 : np);
+   }
+   ParMesh pmesh(MPI_COMM_WORLD, mesh, partitioning);
+   delete [] partitioning;
    mesh.Clear();
    {
       for (int l = 0; l < par_ref_levels; l++)
```
If you want, feel free to merge it here. If not, we can use it as a starting point in a future PR.

File: fem/normal_deriv_restriction.cpp (hunk 15 above)
Comment (v-dobrev) [reply fem/normal_deriv_restriction.cpp:333]: Yes, the other place is here: https://github.com/mfem/mfem/blob/88a79f77889cdf59a306959157a3ea40b2e6360d/fem/prestriction.cpp#L317

It seems natural to do the exchange in `PABilinearFormExtension::Mult` and pass the `ParGridFunction` as input to two new methods that use it to compute `int_face_X` and `int_face_dXdn`. I'm not sure if this will have noticeable impact on performance, so it may not be worth it. What do you think?

**Review by v-dobrev — APPROVED** [review]
Approving and leaving the choice of including the following suggestions to the authors:
* https://github.com/mfem/mfem/pull/3904#discussion_r1562113877, and/or
* https://github.com/mfem/mfem/pull/3904#discussion_r1562081732

File: fem/normal_deriv_restriction.cpp (hunk 15 above)
Comment (pazner) [reply fem/normal_deriv_restriction.cpp]: OK, I implemented something in 936d12883 (and minor change in 54c186eec).

There is a new helper function called `GetLVectorFaceNbrData`. This function checks if the input `Vector&` is actually a ParGridFunction. If so, and if the face-neighbor data is not empty, it just returns an alias to the existing face-neighbor data, without exchanging.

Otherwise, it does the exchange and returns the vector.

This way, if ParGridFunctions are passed into to `ParL2FaceRestriction` or `L2NormalDerivativeFaceRestriction`, the face-neighbor data can be retrieved without communication.

`PABilinearFormExtension::Mult` creates a temporary ParGridFunction (aliasing the input L-vector `x`), which is used to cache the face-neighbor data.

File: fem/bilinearform_ext.cpp (hunk 17 above)
Comment (pazner) [reply fem/bilinearform_ext.cpp:610]: I'd prefer to do a separate PR (since in my testing it looks like this effects not just DG, but the other examples too) — either you can propose the new PR, or I can do it based on your branch, whatever you prefer.

**Comment by tzanio** [comment]
Can we please add an item about this in `CHANGELOG`?

**Comment by pazner** [comment]
> Can we please add an item about this in `CHANGELOG`?

Done!

Thanks for the helpful suggestions @v-dobrev!

File: fem/prestriction.cpp [hunk 23]
```diff
@@ -329,7 +325,7 @@ void ParL2FaceRestriction::DoubleValuedConformingMult(
    auto d_indices1 = scatter_indices1.Read();
    auto d_indices2 = scatter_indices2.Read();
    auto d_x = Reshape(x.Read(), t?vd:ndofs, t?ndofs:vd);
-   auto d_x_shared = Reshape(x_gf.FaceNbrData().Read(),
+   auto d_x_shared = Reshape(face_nbr_data.Read(),
```
Comment (v-dobrev) [inline fem/prestriction.cpp:328]: Since `face_nbr_data` is an alias (when this method is called with a `ParGridFunction` argument `x`) it will be good to sync it to its base to avoid repeated host -> device transfers. The same should be done in the other places where `GetLVectorFaceNbrData` is used. The sync can be done right here after the call to `Read()` -- one needs to get a pointer to the base `Vector` (or `NULL` when `face_nbr_data` is not an alias) from `GetLVectorFaceNbrData` to be able to do the sync easily.

Comment (v-dobrev) [reply fem/prestriction.cpp:328]: A simpler solution maybe to call `.Read()` for the base `Vector` in `GetLVectorFaceNbrData` before creating the alias.

Comment (pazner) [reply fem/prestriction.cpp:328]: OK, I did this in 94906f661 if I understood your suggestion correctly.

Comment (v-dobrev) [reply fem/prestriction.cpp:328]: Yes, that's what I was thinking.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by pazner** [comment]
Thank you for this great contribution @arotem3!!
